runtime/compiler/control/J9Recompilation.cpp

```diff
@@ -779,4 +772,98 @@ J9::Recompilation::resetPersistentProfileInfo(TR_PersistentMethodInfo *methodInf
    methodInfo->_recentProfileInfo = NULL;
    methodInfo->_bestProfileInfo = NULL;
    }
+
+/**
+ * Serialize a TR_PersistentProfileInfo struct
+ *
+ * @param [in] profileInfo ref to TR_PersistentProfileInfo to be serialized
+ * @param [out] profileInfoStr ref to string which points to the serialized TR_PersistentProfileInfo
+ */
+void
+J9::Recompilation::serializePersistentProfileInfo(TR_PersistentProfileInfo &profileInfo, std::string &profileInfoStr)
+   {
+   uint32_t size = profileInfo.getSizeForSerialization();
+   uint8_t * buffer = (uint8_t *)TR_Memory::jitPersistentAlloc(size);
+   uint8_t * origBuffer = buffer; // buffer gets updated after the serialization
+   profileInfo.serialize(buffer);
+   profileInfoStr.append((char *)origBuffer, size);
```

Comment:
> I would size for the string appropriately and then serialize directly into the string's buffer.

I couldn't find any "safe" way to modify the string buffer directly. Modifying the content by accessing the string buffer directly through `string::c_str()` and `string::data()` is an undefined behavior. So better to avoid doing this.

> If that's not an option, could we use JIT scratch memory instead of persistent memory? The answer depends on whether this serialization is performed during a compilation.

This is being done on the client side during a compilation. I think I can replace `TR_Memory::jitPersistentAlloc` with `TR_Memory::allocateHeapMemory()`.

Reply:
Fixed it.